### PR TITLE
Experiment with flakes

### PR DESCRIPTION
**Pull request by madjar**
I was curious to see what would be required to make home-manager work with
flakes, so I took a stab at it. I'm posting this here for those who are curious
about it.

It works surprisingly well. It delegate all the issues of finding nixpkgs and the
home-manager source tree to nix itself, so a lot of the `update-home` script can
be simplified.

To use it, create a new flake that provide a `homeConfiguration`, built using
`home-manager.lib.mkHome`, like so:

```
{
  edition = 201909;

  inputs.home-manager.uri = "/Users/georgesdubus/dev/perso/home-manager";

  outputs = { self, nixpkgs, home-manager }: {

    homeConfiguration = home-manager.lib.mkHome {
      imports = [ ./home.nix ];
      home.homeDirectory = "/Users/georgesdubus";
      home.username = "georgesdubus";
    };
  };
}

```
Then build the activation package with
`nix build .#homeConfiguration.activationPackage`. Or run the home-manager
script like so:
`nix run /Users/georgesdubus/dev/perso/home-manager -c home-manager build -f .`.
This assumes that you have a flake-enabled nix that you can get with `nix run nixpkgs.nixFlakes`.

Outstanding issues include:
- I've hardcoded it to darwin because that's the plaform I develop on, but it
  should be too much of a problem to fix that.
- Because the evaluation is pure, there is no way to access the $USER and $HOME
  environment variables, which means they need to be passed manually. I'm
  guessing it could be made nicer by having the wrapping script get them and
  pass them.
- Since the flake takes care of providing nixpkgs, it's not possible to provide
  configuration for it in `options.nixpkgs`. I believe this is still an open
  question for the flake RFC, so wait and see.
- Channels are not available, and nixos modules are not yet exposed from the
  nixpkgs flake, so the source of nixpkgs is passed manually to find the
  `assertions.nix` and `meta.nix` modules.
- I'm guessing we'll want to get closer to the behaviour of nixos-rebuild, once
  it's been finalized (multiple attributes in the `homeConfiguration` output,
  that kind of things)
- The manual didn't work out the box. I disabled it rather than looking into it.

### CODE REVIEW

**Comment by rycee** [comment]
That's awesome! I haven't had time to look into the PR yet but I appreciate you taking the time to get things going. I hope to find the time to review and comment properly as soon as possible.

**Comment by teto** [comment]
that's fantastic ! I will give it a shot as well.

File: doc/default.nix [hunk 1]
```diff
@@ -29,7 +29,7 @@ let
 
   hmModulesDocs = nmd.buildModulesDocs {
     modules =
-      import ../modules/modules.nix { inherit lib pkgs; }
+      import ../modules/modules.nix { inherit lib pkgs;}
```
Comment (teto) [inline doc/default.nix]: unnecessary change, is it a formatter ?

**Comment by teto** [comment]
would you mind fixing the `I've hardcoded it to darwin because that's the plaform I develop on, but it
should be too much of a problem to fix that.` please ?

File: doc/default.nix (hunk 1 above)
Comment (madjar) [reply doc/default.nix]: Possibly. I removed it :)

**Comment by madjar** [comment]
@teto I looked a bit into unhardcoding the `system`, but it's a bit trickier than I expected. I'll explain.

So, the basic idea would be to use the same pattern Eelco uses (https://github.com/edolstra/nixpkgs/blob/master/flake.nix#L15-L17).
That would look like:
```nix
let
  systems = [ "x86_64-linux" "i686-linux" "x86_64-darwin" "aarch64-linux" ];
  forAllSystems = f: lib.genAttrs systems (system: f system);
in {
  packages = forAllSystems (system: {home-manager = nixpkgs.legacyPackages.${system}.callPackage ./home-manager {};});
}
```

The tricky part is that, now, I can't just pass the `nixpkgs` I want as `_module.args.pkgs`. Instead, I would need to fix `
home-manager/modules/misc/nixpkgs.nix` to be aware of how this new nixpkgs works, and get the proper package set for the right system (which may need to be defined manually).

It's definitely doable, but a bit of work, and I'm really curious to see what comes after `legacyPackages`, and how this same issue get solved for nixos systems, before I start putting too much time into this.

**Comment by teto** [comment]
would https://github.com/rycee/home-manager/pull/993 help ? 

i completely understand not putting too much work as flakes are a moving target. Likewise I really want to use it as it solves many problems but right now it creates too many as well xD

**Comment by madjar** [comment]
Oh, #993 might very well help, in that it would allow to define `_module.args.pkgs` in another module (one that knows about flakes), completely side-stepping `modules/misc/nixpkgs.nix`!

I'll look into it, it might actually be easy.

**Comment by cole-h** [comment]
Now that flake support have been merged into master, I believe this can be closed? :)

**Comment by teto** [comment]
This PR contains some changes not present in master like the defaultPackage etc so it could be interested to rebase it instead.

**Comment by madjar** [comment]
All the changes in here have been incorporated in master in some way or other (except the changes to the `home-manager` script, but I'm not sure they're needed).

I have a nice home-manager flake setup that uses home-manager's master now, so I guess everything is good :)
I'll have to document it at some point, but as a teaser, I switch to a new configuration by running `nix run` thanks to:
```
defaultApp.x86_64-darwin = {
      type = "app";
      program = "${self.home.activationPackage}/activate";
    };
```

**Comment by Mic92** [comment]
friendly ping.

**Comment by madjar** [comment]
Well, flake support in master is great and works well, so let's close this!
